### PR TITLE
IBP-5874 Implement Two Factor Authentication

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

File: src/main/web/src/pages/login.html [hunk 1]
```diff
@@ -103,6 +103,24 @@
...
+		<p>Please enter the verification code sent to your registered email:</p>
+		<div class="login-form-control">
+			<input class="js-otp-code login-otp-input login-flat-control" id="otpCode" name="otpCode" type=text" th:maxlength="${otpCodeLength}" autocomplete="off"></input>
+		</div>
+		<div class="js-login-otp-error login-otp-valid">
+			<span class="fa login-warning-icon">&#xf12a;</span>
+			<span class="js-login-otp-error-text"></span>
+		</div>
+		<button class="js-verify-otp-submit login-submit login-flat-control login-clickable" formnovalidate="" type="button">
+			<span class="login-submit-label">Verify OTP</span>
```
Comment (clarysabel) [inline src/main/web/src/pages/login.html:119]: @abatac use Confirm or Verify only

**Comment by clarysabel** [comment]
@abatac 
I was trying the new feature locally. For an instance with 2FA enabled, I created a new user with 2FA disabled. 
I received the email to change the password, changed it and BMS required to login, I enter user and password and the verification code was required, even when in the database the field was set in 0. 
Could you please check this? 
CC @nahuel-soldevilla

**Review by nahuel-soldevilla — COMMENTED** [review]
Looks good. 
I guess it makes sense to still keep this logic here in Workbench auth controller instead of bmsapi, otherwise we'll need to make too many request to bmsapi, probably duplicating the same logic + a bunch of network calls.

**Comment by abatac** [comment]
> @abatac I was trying the new feature locally. For an instance with 2FA enabled, I created a new user with 2FA disabled. I received the email to change the password, changed it and BMS required to login, I enter user and password and the verification code was required, even when in the database the field was set in 0. Could you please check this? CC @nahuel-soldevilla

I'll check this.

File: src/main/web/src/pages/login.html (hunk 1 above)
Comment (abatac) [reply src/main/web/src/pages/login.html]: Done

File: src/main/java/org/generationcp/ibpworkbench/security/WorkbenchEmailSenderService.java [hunk 2]
```diff
@@ -76,6 +75,89 @@ public class WorkbenchEmailSenderService {
...
+
+			// Send the message
+			this.mailSender.send(mimeMessage);
+
+		} catch (final IOException e) {
+			WorkbenchEmailSenderService.LOG.error(e.getMessage(), e);
+		}
+	}
+
+	public void doSendOneTimePasswordRequestForUnknownDevice(final WorkbenchUser user, final Integer otpCode,
```
[the remark this answers is not in the document]
Comment (abatac) [reply src/main/java/org/generationcp/ibpworkbench/security/WorkbenchEmailSenderService.java]: Done

File: src/main/java/org/generationcp/ibpworkbench/security/WorkbenchEmailSenderService.java [hunk 3]
```diff
@@ -76,6 +75,89 @@ public class WorkbenchEmailSenderService {
...
+			message.setText(htmlContent, true);
+
+			// add BMS logo
+			message.addInline(WorkbenchEmailSenderService.BMS_LOGO_LOC, this.retrieveLogoImage(), "image/png");
+
+			// Send the message
+			this.mailSender.send(mimeMessage);
+
+		} catch (final IOException e) {
+			WorkbenchEmailSenderService.LOG.error(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (abatac) [reply src/main/java/org/generationcp/ibpworkbench/security/WorkbenchEmailSenderService.java:117]: Done

File: src/main/java/org/generationcp/ibpworkbench/controller/AuthenticationController.java [hunk 4]
```diff
@@ -159,59 +205,166 @@ private void populateCommomModelAttributes(final Model model) {
...
-			if (result.hasErrors()) {
-				this.generateErrors(result, out);
-				return new ResponseEntity<>(out, isSuccess);
-			}
+	private Integer getNumberOfOTPVerificationAtemptPerUser(final String userName) {
+		try {
+			return this.otpVerificationAttemptCache.get(userName);
+		} catch (final Exception e) {
+			// If number of attempt does not yet exist per user, just return 1 (first attempt)
+			return 1;
```
[the remark this answers is not in the document]
Comment (abatac) [reply src/main/java/org/generationcp/ibpworkbench/controller/AuthenticationController.java]: I checked and it doesn't happen. I already removed this unnecessary method, and changed the initialization value to 1.

**Comment by abatac** [comment]
> > @abatac I was trying the new feature locally. For an instance with 2FA enabled, I created a new user with 2FA disabled. I received the email to change the password, changed it and BMS required to login, I enter user and password and the verification code was required, even when in the database the field was set in 0. Could you please check this? CC @nahuel-soldevilla
> 
> I'll check this.

This is done. If the user is not enabled for 2FA and logs in for the first time (no existing user devices yet), OTP verification should not happen.

**Review by nahuel-soldevilla — APPROVED** [review]

**Pull request merged**